<a href="https://colab.research.google.com/github/seandaza/Python-The-Fundamentals/blob/master/Loops.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# I. For Loops

> Los iteradores del tipo `For` son estructuras de control que permiten recorrer estructuras repetitivas de datos de forma simple y ágil, pudiendo configurar el inicio del recorrido y el numero de bucles a ejecutar. Veamos algunos ejemplos:


>  <img src = "https://sharpiness.files.wordpress.com/2019/02/dominoes-falling_0.jpg?w=551&h=810" alt = "ejemplo" align = "center" />







##Sección 1: Recorrer una Cadena

Recordemos que una cadena es un conjunto de caracteres seguidos unos tras de otros, incluyendo los espacios. A continuación declararemos una variable `S` que es la que almacenará mi cadena; de tal manera que, el recorrido se hara sobre esa variable que contiene la cadena. Por otro lado, hay que indicarle que haga algo por cada caracter en la cadena, en este caso, imprimiremos la letra sobre la que se esta recorriendo.

###Ejemplo 1


In [ ]:
#Declaremos una cadena aleatoria
S = "Hola a Todos!"
#Definamos el recorrido
for char in S:
  print(char)


H
o
l
a
 
a
 
T
o
d
o
s
!


###Analisis

> Notemos que por cada caracter de la cadena S que recorre, se imprime el mismo. Ahora bien, ¿Cómo hacenos si queremos solo imprimir las vocales de esta cadena? . Para ello, declaremos una variable denominada 'vocales' que almacena las vocales en mayuscula y minuscula en formato de cadena y condicionemos nuestra busqueda por recorrido pidiendo que solo se impirman los caracteres que esten tambien en la cadena 'vocales'. Veamos:




###Ejemplo 2

In [ ]:
#Declaramos nuestra cadena
S = "Hola a todos!"

#Declaramos en una cadena las vocales (minusculas y mayusculas)
vocales = 'aeiouAEIOU'

#Recorremos y condicionamos nuestra busqueda
for char in S:
  if char in vocales:
    print(char)

o
a
a
o
o


###Pregunta


> ¿Cómo hacemos si necesitamos Contar el numero de vocales que contiene una cadena?. Para resolver este problema, haremos uso de **Acumuladores**. Veamos en la siguiente seccion de que se trata.



##Sección 2: Acumuladores


> Un Acumulador es una variable inicializada en el valor cero de tal manera que me permita ir cargando las veces que una condición se cumpla por recorrido del bucle. Es decir,  un acumulador, es básicamente un **contador** que se va incrementando gradualmente en la medida que el loop vaya haciendo match en el recorrido sujeto la condición a la que se le imponga.

> Para continuar con el ejemplo anterior, incorporemos un acumulador en el for loop, para ir contando las veces que se identifica una vocal en el bucle. Asi:





###Ejemplo 1

In [ ]:
# Cadena sobre la que se itera
S ="Hola a Todos!"

# Cadena que contiene a las vocales (en version mayuscula y minuscula)
vocales = 'aeiouAEIOU'

# Declaremos nuestro contador como una variable que lleva el mismo nombre (contador) e inicialicemmosla con un valor inicial dede 0
# y efectuemos el bucle condiconado:
contador = 0
for char in S:
  if char in vocales:
    contador = contador + 1
print(contador)



5


###Observaciones y Preguntas


> 1. Note que asi tal cual nuestro código, nos imprime al final el número total de vocales como queriamos.

> 2. Note que la última linea `print(contador)` esta alineada al `for` del inicio del código. Prueba alineando la ultima linea `print(contador)` al `if`, oprimiendo la tecla 'TAB' de tu teclado con el cursor al inicio de la linea y ejecuta el código. ¿Qué observas en la salida de la consola? Explica lo que se muestra por consola.

> 3. Alinea ahora la última linea `print(contador)` al nivel de la linea penúltima. ¿Qué observas en la salida de la consola? Explicalo.

> 4. Cambia la penúltima linea: `contador = contador + 1` por la expresión: 
  `contador +=1` y comprueba que en escencia, efectua lo mismo, pero es una forma mas elegante de escribir el código. 









###Ejemplo2



> En el ejemplo 2 de la sección anterior, construimos un for loop que nos imprime las vocales que contiene determianda cadena, pero por cada vocal hacía una impresión. Nuestro objetivo ahora es imprimir una cadena que contenga las vocales que aparecen en la cadena dada, a efectos de solo hacer una única impresión. Para ello, haremos uso nuevamente de un acumulador. Veamos:




In [ ]:
# Declaramos una cadena aleatoria
S = "Happy Anniversary"

#Declaramos el conjunto de las vocales en version mayuscula y minuscula
vocales = 'aeiouAEIOU'

# Declaramos nuestro acumulador inicializado como una cadena vacía, recorremos la cadena
# y le vamos Concatenando a nuestro acumulador, las letras que sean vocales
acumulador = ''
for char in S:
  if char in vocales:
    acumulador += char
print(acumulador)



###Observaciones

> Note que nuestro acumulador es una cadena vacía y que nuestra opcion aqui es ir Concatenandole a esta cadena, las vocales que se vayan identidicando en la cadena a detallar.

> Note que para Concatenar también se usa el operador `+`. En este caso no hay problema porque estamos operando datos del mismo tipo.





###Ejercicios

####Ejercicio 1

> *Necesitamos* verificar si una cadena cualquiera contiene almenos una vocal, en este caso, nos debe imprimir por consola `True`; en caso contrario, debe imprimir por consola `False`.

> Para ello, debemos recorrer la cadena con un `for` y usar una variable `vowel_found` que hará el papel de un switch o interruptor que por defecto estará "apagado" o en modo `False`. Si en el recorrido del `for` la cadena hace match con una vocal, entonces nuestro interuptor se "encenderá" y por lo tanto nuestra variable `vowel_found` deberá actualizarse.

> ![Texto alternativo](	https://stem.guide/wp-content/uploads/2018/08/shutterstock_306139223.jpg)

> Para ello, su objetivo será completar el código y comprobar que retorna lo que se pide:










In [ ]:
    vowel_found = False
    for char in s:
        if char in 'aeiouAEIOU':
          #escribe el codigo faltante aquí  

    return vowel_found

####Ejercicio 2


> Sean `S1` y `S2` dos cadenas de texto diferentes cualesquiera. Escriba un código que me permita comparar ambas cadenas y me retorne las letras que tienen en común ambas cadenas. por ejemplo, si `S1 = 'abc'` y `S2 = 'ad'`, entonces nuestro código nos debería mostrar por consola `'a'`, que correspondería a la letra que tienen en común ambas cadenas. Si por ejemplo `s1 = 'abb'` y `S2 = 'ab'`, entonces nuestro código nos debería mostrar como salida por consola `'abb'`.







In [ ]:
# Construye tu código siguiendo los siguientes pasos:

# 1. Declara las Cadenas S1 y S2

S1 = 'abb'
S2 = 'ab'

# 2. Declara un acumulador


# 3. Declara un For Loop de elementos de una cadena que vaya comparandolos con los elementos de la otra cadena
#    y condiciona la comparación. En caso de que hagan match, guarda la letra en el acumulador.




# 4. Imprime lo que quieres mostrar





#### Advertencia!!


> A continuación puedes ver la respuesta si y solamente sí trabajaste en tus propias ideas para generarla y poder retroalimentar tu esfuerzo.



In [ ]:
#@title Respuesta { display-mode: "form" }

S1 = 'abb'
S2 = 'ab'

acum = ''
for elm in S1:
  if elm in S2:
    acum+= elm
print(acum)
    

# II. While Loops

> Un bucle while permite repetir la ejecución de un grupo de instrucciones mientras se cumpla una condición (es decir, mientras la condición tenga el valor True).





<img src = "https://m.media-amazon.com/images/I/61E3a5EiOsL._AC_SL1000_.jpg" width = "500" height = "500" alt = "ejemplo" align = "center" />



	



## Estructura

La sintaxis del bucle while es la siguiente:

In [ ]:
while condicion:
    cuerpo del bucle

La ejecución de esta estructura de control while es la siguiente:

Python evalúa la condición:

1. si el resultado es True se ejecuta el cuerpo del bucle. Una vez ejecutado el cuerpo del bucle, se repite el proceso (se evalúa de nuevo la condición y, si es cierta, se ejecuta de nuevo el cuerpo del bucle) una y otra vez mientras la condición sea cierta.

2. si el resultado es False, el cuerpo del bucle no se ejecuta y continúa la ejecución del resto del programa.
La variable o las variables que aparezcan en la condición se suelen llamar variables de control. 

Las variables de control deben definirse antes del bucle while y modificarse en el bucle while.

## Sección 1: Condición de Tipo Numérica

> En primera instancia realicemos el bucle sobre una condición de tipo numérico; es decir, se puede verificar algún tipo de relación comparativa entre valores numéricos.



###Ejemplo 1


> Guardemos en una variable un valor numérico, por cada ciclo, vamos a duplicar ese numero y actualizamos el valor de la variable al mismo. Si el valor de la variable final es menor que 100, repetimos el ciclo hasta que se cumpla esta relacion de orden. Veamos:



In [31]:
num = 10
while num < 100:
  num = num * 2
  print(num)

20
40
80
160


###Análisis


> Note que `80` es el último valor de la variable `num` que entra en el ciclo e imprime su duplo `160`



###Ejemplo 2

> Veamos un caso en el que la condición NUNCA se cumple y por lo tanto nunca se ejecuta nada. Por lo tanto tampoco va a imprimir nada por consola.



In [32]:
num = 100
while num < 100:
  num = num * 2
  print(num)


### Análisis

> La Condición sobre la que se ejecuta un `While`, debe ser CIERTA para que el flujo lógico ejecute todo lo que está dentro de él.



###Ejemplo 3

> Si la condición sobre la que se construye el  `While` es SIEMPRE cierta, podemos incurrir en Bucles Infinitos y podriamos consumir toda la memoria en este proceso. Debemos ser cuidadosos para no incurrir en estos casos. Veamos un ejemplo:



In [ ]:
while 1 > 0:
  print("hola")

###Análisis

> Como sabemos, 1 siempre es mayor que 0, por lo tanto la condición sobre la que construímos nuestro `While`, es SIEMPRE cierta; por lo tanto, en este caso, nos va a imprimir `'hola'` de forma indefinida hasta que paremos el proceso de manera forzosa.



##Sección 2: Condición de Tipo Booleana

> Podemos exigir que nuestra condición no sea extrictamente una relación numérica, sino también podemos usar cualquier compardor lógico.



###Ejemplo 1

> Supongamos que tenemos una cadena de caracteres `S` y queremos que nos vaya imprimiendo letra a letra  de esta cadena, mientras la letra sea una consonante y que pare cuando encuentre una vocal. Entonces:



In [42]:
# Supongamos la Cadena es 
S = 'psicologia'

i=0
while not (S[i] in 'aeiouAEIOU'):
  print(S[i])
  i += 1



P
s


### Para Pensar ...

> 1. Explica la salida anterior con tus palabras.

> 2. Ejecuta el mismo código anterior, pero asume ahora que tu cadena es `S = 'agua'` y explica.

> 3. Ejecuta el mismo código anterior, pero asume ahora que tu cadena es `S = 'xyz'` y explica.






En el último caso, lee que error te muestra la consola y trata de entenderlo. Para enmendar ese error, prueba el siguiente código, detalla los adicionales y explica que ocurrió:

In [ ]:
S = 'xyz'
i = 0
while i < len(S) and not (S[i] in 'aeiouAEIOU'):
    print(S[i])
    i = i + 1

###Ejercicio

> Haciendo uso de un `while`, escriba de forma descendiente los primeros 10 numeros naturales; es decir, `10,9,8,7,6,5,4,3,2,1`


In [ ]:
# Escribe tu código aquí!
